xmtp_mls/src/groups/sync.rs

```diff
@@ -0,0 +1,771 @@
+            Err(GroupError::ReceiveErrors(receive_errors))
+        }
+    }
+
+    pub(super) async fn receive<'a>(&self, conn: &'a DbConnection<'a>) -> Result<(), GroupError> {
+        let messages = self
+            .client
+            .query_group_messages(&self.group_id, conn)
+            .await?;
+
+        self.process_messages(messages, conn)?;
+
+        Ok(())
+    }
+
+    fn save_transcript_message(
+        &self,
+        conn: &DbConnection,
+        maybe_validated_commit: Option<ValidatedCommit>,
+        timestamp_ns: u64,
+    ) -> Result<Option<StoredGroupMessage>, MessageProcessingError> {
+        let mut transcript_message = None;
+        if let Some(validated_commit) = maybe_validated_commit {
+            // If there are no members added or removed, don't write a transcript message
+            if validated_commit.members_added.is_empty()
+                && validated_commit.members_removed.is_empty()
+            {
+                return Ok(None);
+            }
+            let sender_installation_id = validated_commit.actor_installation_id();
+            let sender_account_address = validated_commit.actor_account_address();
+            let payload: GroupMembershipChanges = validated_commit.into();
+            let encoded_payload = GroupMembershipChangeCodec::encode(payload)?;
+            let mut encoded_payload_bytes = Vec::new();
+            encoded_payload.encode(&mut encoded_payload_bytes)?;
+            let group_id = self.group_id.as_slice();
+            let message_id =
+                get_message_id(encoded_payload_bytes.as_slice(), group_id, timestamp_ns);
+            let msg = StoredGroupMessage {
+                id: message_id,
+                group_id: group_id.to_vec(),
+                decrypted_message_bytes: encoded_payload_bytes.to_vec(),
+                sent_at_ns: timestamp_ns as i64,
+                kind: GroupMessageKind::MembershipChange,
+                sender_installation_id,
+                sender_account_address,
+            };
+
+            msg.store(conn)?;
+            transcript_message = Some(msg);
+        }
+
+        Ok(transcript_message)
+    }
+
+    pub(super) async fn publish_intents<'a>(
+        &self,
+        conn: &'a DbConnection<'a>,
+    ) -> Result<(), GroupError> {
+        let provider = self.client.mls_provider(conn);
+        let mut openmls_group = self.load_mls_group(&provider)?;
+
+        let intents = provider.conn().find_group_intents(
+            self.group_id.clone(),
+            Some(vec![IntentState::ToPublish]),
+            None,
+        )?;
+        let num_intents = intents.len();
+
+        for intent in intents {
+            let result = retry_async!(
+                Retry::default(),
+                (async {
+                    self.get_publish_intent_data(&provider, &mut openmls_group, &intent)
+                        .await
+                })
+            );
+
+            if let Err(err) = result {
```

Comment:
Either set the intent to Error status or increment the attempt count

Reply:
Previously it would continue to the next intent in case of error. Now it will return to make sure that intents get processed in order.